File: catalogue/garde_manger.py
"""Le garde-manger : les rangements physiques, et la matière première dedans.

UN OBJET DE PLUS, PARCE QUE LES DEUX QUI EXISTAIENT NE POUVAIENT PAS LE PORTER.

- `stock.yaml` porte les sorties de CUISINE, indexées sur les types que les
  recettes émettent. Une conserve de maïs n'en est pas une, et lui donner
  `kind: base` la ferait entrer dans le graphe de chaînage.
- `rayons.placard` marque ce qu'on possède TOUJOURS — sel, huile, poivre. C'est
  une appartenance, sans quantité ni endroit : elle dit qu'on a du sel, jamais
  qu'on a quatre boîtes de maïs de 285 g.

Ce module lit le troisième, le valide, et dérive de ses dimensions les deux
seules grandeurs qu'on ne veut pas voir saisies à la main : le volume d'une zone
et le poids qu'elle porte. Une donnée saisie deux fois diverge ; ces deux-là se
calculent.

CE QUE LE MODÈLE NE FAIT PAS. Rien ne décrémente une denrée quand on cuisine, et
le garde-manger reste donc DESCRIPTIF : `provenance()` continue de lire
`rayons.placard`, pas ce fichier. Brancher les deux demande un modèle de
consommation qui n'existe pas encore — c'est au backlog.
"""

from pathlib import Path

import yaml

import chainage as ch

# Ce qui peut abîmer une denrée, et l'attribut de zone qui le contredit. La
# table EST la règle : ajouter une sensibilité, c'est ajouter une ligne ici, et
# le vérificateur la joue sans rien savoir de plus.
AGRESSIONS = {
    "lumiere": ("exposition", "jour", "à la lumière du jour"),
    "humidite": ("hygrometrie", "humide", "dans un endroit humide"),
    "chaleur": ("chaleur", True, "près d'une source de chaleur"),
}

EXPOSITIONS = ("jour", "sombre")
HYGROMETRIES = ("sec", "humide")

# LA FORME CORRIGE LE VOLUME, ET L'ÉCART N'EST PAS ANECDOTIQUE. Les deux plateaux
# d'angle sont des demi-disques : leur surface vaut π/4 de celle du rectangle qui
# les contient, soit 21 % de moins. Les compter en boîte donnait 67 L pour un
# plateau qui en porte 53, et un budget de rangement faux d'un cinquième est un
# budget qui déborde sans prévenir.
FORMES = {"rectangle": 1.0, "demi-lune": 0.785}

# L'état dit COMMENT c'est conditionné, ce qui décide de la barrière que ça
# oppose au monde. Un sachet entamé et une conserve ne se conservent pas pareil,
# et ce n'est pas la denrée qui change — c'est son emballage.
ETATS = ("conserve", "bocal", "sec", "entame", "frais")


# Un id se lit à l'écran, et « pignons-pin » n'est pas du français. Le
# dé-tiretisage vit ici parce que l'alerte est écrite ici : elle doit sortir
# lisible du modèle, pas être recousue par l'interface.
def _nom(ingredient: str) -> str:
    return ingredient.replace("-", " ")


def _label(zone: dict) -> str:
    return zone.get("label") or zone["id"]


# Trois urgences, et AUCUNE DATE. Le relevé n'en porte pas : ni DLC, ni DLUO, ni
# date d'ouverture. Inventer une échéance pour pouvoir compter dessus serait le
# genre de chiffre qui a l'air juste et ne l'est jamais.
#
# Ce que le relevé sait vraiment, c'est le CONDITIONNEMENT et l'ENDROIT. Un
# sachet ouvert n'oppose plus de barrière ; un légume frais ne se garde pas ; une
# denrée rangée là où elle s'abîme se dégrade en ce moment. Ces trois faits
# suffisent à classer, et ils sont vérifiables — on peut aller les regarder.
URGENCES = ("haute", "moyenne", "basse")


def urgence(denree: dict, zone: dict) -> str:
    """À quel point il faut manger ça bientôt.

    `haute` — le frais, et tout ce que sa zone abîme activement.
    `moyenne` — le paquet entamé : la barrière est rompue, l'horloge tourne.
    `basse` — scellé. Une conserve attend des années sans rien demander.
    """
    if denree.get("etat") == "frais":
        return "haute"
    for s in denree.get("sensible") or []:
        if s in AGRESSIONS:
            attr, mauvais, _ = AGRESSIONS[s]
            if zone.get(attr) == mauvais:
                return "haute"
    return "moyenne" if denree.get("etat") == "entame" else "basse"


NATURES = ("legume-cru", "fruit", "herbe", "sec", "gras", "plat", "autre")

# Le défaut d'acidité vient de `conservation.yaml` (`defaut_acidite: basse`) et
# c'est un choix de SÉCURITÉ. Voir l'avertissement botulisme en tête de ce
# fichier-là : le bain-marie ne stérilise que les aliments acides, et sur un
# légume nature en bocal à température ambiante il crée exactement le milieu
# anaérobie peu acide où prolifère C. botulinum. Une denrée est donc présumée
# PEU ACIDE tant qu'elle ne dit pas le contraire.
ACIDITE_DEFAUT = "basse"


def charger(chemin: Path) -> dict:
    """`{'zones': [...], 'denrees': [...]}`, jamais `None` sur un fichier vide."""
    data = yaml.safe_load(chemin.read_text()) or {}
    return {"zones": data.get("zones") or [], "denrees": data.get("denrees") or []}


def conservations(denree: dict, methodes: list, capacites: set) -> list:
    """Ce qu'on peut faire de cette denrée pour arrêter son horloge.

    LA SECONDE RÉPONSE AU GASPILLAGE, et elle existait avant ce fichier :
    `conservation.yaml` la porte depuis le prototype. Une denrée qui court ne se
    rattrape pas seulement en la cuisinant ce soir — elle se rattrape en la
    transformant, à condition d'avoir le kit et le geste.

    Rend une liste de `{id, label, acquis, fenetre, manque, noeud}`, y compris
    les méthodes VERROUILLÉES : c'est le principe du nœud de compétence de #10.
    Une méthode qu'on ne possède pas n'est pas une erreur à taire, c'est ce qu'il
    faudrait acquérir — et le dire est la moitié de l'intérêt.

    TROIS FILTRES, ET LE PREMIER TUE DES GENS S'IL SAUTE :

    1. `exige_acidite: haute` contre l'acidité de la denrée. Le bain-marie sur un
       aliment peu acide est la faute dangereuse de ce domaine.
    2. `applique_a` contre la nature. On ne lacto-fermente pas de la farine.
    3. `conserve_mal` sur la denrée, pour ce que le modèle général rate — la
       pomme de terre crue au congélateur.
    """
    nature = denree.get("nature", "autre")
    acide = denree.get("acidite", ACIDITE_DEFAUT)
    exclues = set(denree.get("conserve_mal") or [])
    sorties = []
    for m in methodes:
        if m["id"] in exclues:
            continue
        if m.get("exige_acidite") == "haute" and acide != "haute":
            continue
        vise = m.get("applique_a")
        if vise and nature not in vise:
            continue
        besoin = m.get("needs")
        noeud = m.get("noeud_competence") or {}
        sorties.append({
            "id": m["id"],
            "label": m["label"],
            "acquis": besoin is None or besoin in capacites,
            "fenetre": _fenetre(m.get("fenetre")),
            "manque": noeud.get("kit_manquant") or besoin,
            "noeud": noeud.get("titre"),
            # UNE MÉTHODE TAILLÉE POUR CETTE MATIÈRE, par opposition à une
            # méthode passe-partout. `vise` sans `plat` veut dire que la méthode
            # a été écrite pour de la matière première : la lacto-fermentation et
            # le séchage, pas le sous-vide qui marche sur à peu près tout.
            #
            # La distinction sert à l'écran. Le sous-vide est le premier verrou
            # de TOUTES les denrées, et l'afficher partout écrivait treize fois
            # la même phrase — un conseil qu'on lit treize fois est un conseil
            # qu'on ne lit plus. Ne se propose donc que ce qui est spécifique.
            "specifique": bool(vise) and "plat" not in vise,
        })
    return sorties


def _fenetre(f) -> str:
    """« 3 mois », « 4 jours », ou « ×2,5 » — la forme que l'écran lira.

    Un multiplicateur n'est pas une durée : le sous-vide RALLONGE le froid, il ne
    donne pas une fenêtre à lui. Les rendre pareils ferait lire « 2,5 jours » là
    où il faut lire « deux fois et demie plus longtemps ».
    """
    if not f:
        return ""
    if f.get("multiplicateur"):
        return f"×{f['multiplicateur']}".replace(".", ",")
    return f"{f['valeur']} {f['unite']}"


def volume_litres(zone: dict):
    """Le volume utile d'une zone, en litres, ou `None`.

    `None` DÈS QU'UNE COTE MANQUE, et c'est voulu : le tiroir à épices n'a pas
    été mesuré, et lui inventer un volume en supposant une cote le ferait entrer
    dans un total qui serait faux sans le dire. Une zone à hauteur libre (le
    dessus de l'étagère) n'a pas de volume non plus — c'est une surface.
    """
    d = zone.get("dimensions") or {}
    cotes = [d.get("largeur_cm"), d.get("profondeur_cm"), d.get("hauteur_cm")]
    if any(c is None for c in cotes):
        return None
    coef = FORMES.get(zone.get("forme", "rectangle"), 1.0)
    return round(cotes[0] * cotes[1] * cotes[2] * zone.get("niveaux", 1) * coef / 1000, 1)


def poids_g(denree: dict):
    """La matière d'une ligne, en grammes, ou `None` si elle n'est pas pesée.

    On ne convertit pas : une ligne en `ml` ou en `pièce` ne s'additionne pas à
    des grammes, et faire semblant produirait un total qui a l'air d'un poids.
    """
    q = denree.get("par_unite")
    if not q or q.get("unit") != "g" or q.get("amount") is None:
        return None
    return q["amount"] * denree.get("unites", 1)


def vocabulaire(rayons: dict) -> set:
    """Tous les ids d'ingrédients connus — les rayons ET le placard."""
    return {i for v in rayons["rayons"].values() for i in v} | set(rayons.get("placard") or [])


def alertes_rangement(gm: dict) -> list:
    """Les erreurs de RANGEMENT — et rien d'autre.

    DEUX NATURES D'AVERTISSEMENT COHABITAIENT, ET LES MÉLANGER LES ABÎMAIT TOUTES
    LES DEUX. « Le tiroir à épices n'a pas de cotes » parle du fichier et
    s'adresse à qui le tient ; « les pignons de pin sont à la lumière » parle de
    la cuisine et s'adresse à qui l'habite. Envoyées ensemble à l'écran, les
    secondes se noyaient dans les premières — et l'app affichait « à déplacer »
    au-dessus d'une remarque qu'aucun geste dans la cuisine ne peut résoudre.

    Cette fonction ne rend que ce qui se corrige en déplaçant quelque chose.
    `verifier_garde_manger` rend tout, parce que le vérificateur, lui, sert à
    tenir le fichier.
    """
    zones = {z["id"]: z for z in gm["zones"]}
    alertes = []

    par_zone = {}
    for d in gm["denrees"]:
        if d.get("zone") in zones:
            par_zone.setdefault(d["zone"], []).append(d)

    # UNE ALERTE PAR GESTE, ET C'EST CE QUI DÉCIDE DU REGROUPEMENT.
    #
    # Une ligne par (denrée × agression) donnait sept alertes pour trois
    # problèmes : les pignons de pin comptaient double — la lumière ET la
    # bouilloire — alors qu'un seul déplacement les règle, et le sous-évier
    # répétait quatre fois « dans un endroit humide » sous quatre légumes qu'on
    # sortira du même mouvement. Sept lignes ne se lisent pas ; trois, si.
    #
    # On groupe donc par zone ET par diagnostic identique : ce qui se corrige
    # ensemble se dit ensemble.
    for zid, dedans in par_zone.items():
        z = zones[zid]
        groupes = {}
        for d in dedans:
            dits = tuple(AGRESSIONS[s][2] for s in d.get("sensible") or []
                         if s in AGRESSIONS and z.get(AGRESSIONS[s][0]) == AGRESSIONS[s][1])
            if dits:
                groupes.setdefault(dits, []).append(_nom(d["ingredient"]))
        for dits, noms in groupes.items():
            alertes.append(f"{', '.join(noms)} — {' et '.join(dits)} ({_label(z)})")

        presents = {d["ingredient"] for d in dedans}
        for d in dedans:
            fautifs = sorted(_nom(f) for f in presents & set(d.get("incompatibles") or []))
            if fautifs:
                alertes.append(f"{_nom(d['ingredient'])} — ne devrait pas voisiner "
                               f"avec {', '.join(fautifs)} ({_label(z)})")
    return alertes


def verifier_garde_manger(gm: dict, rayons: dict) -> tuple:
    """Erreurs et avertissements du garde-manger, pour le vérificateur.

    LES ERREURS SONT DES INCOHÉRENCES DE DONNÉES : une zone en double, un espace
    inconnu, une denrée rattachée à un rangement qui n'existe pas, un ingrédient
    hors vocabulaire. Toutes se corrigent en éditant le fichier.

    LES AVERTISSEMENTS MÊLENT LES DEUX MONDES, et c'est voulu ICI : qui lance le
    vérificateur veut tout voir, la donnée incomplète comme le sachet mal rangé.
    C'est l'EXPORT qui trie — voir `alertes_rangement`.
    """
    err, warn = [], []
    zones = {}

    for z in gm["zones"]:
        zid = z.get("id", "?")
        if zid in zones:
            err.append(f"zone « {zid} » déclarée deux fois")
        zones[zid] = z
        if z.get("espace") not in ch.ESPACES:
            err.append(f"zone « {zid} » : espace « {z.get('espace')} » inconnu "
                       f"— attendu parmi {list(ch.ESPACES)}")
        if z.get("exposition") not in EXPOSITIONS:
            err.append(f"zone « {zid} » : exposition « {z.get('exposition')} » inconnue "
                       f"— attendu parmi {list(EXPOSITIONS)}")
        if z.get("hygrometrie") not in HYGROMETRIES:
            err.append(f"zone « {zid} » : hygrométrie « {z.get('hygrometrie')} » inconnue "
                       f"— attendu parmi {list(HYGROMETRIES)}")
        if not z.get("niveaux"):
            err.append(f"zone « {zid} » : il faut au moins un niveau")
        if z.get("forme", "rectangle") not in FORMES:
            err.append(f"zone « {zid} » : forme « {z.get('forme')} » inconnue "
                       f"— attendu parmi {sorted(FORMES)}")
        if volume_litres(z) is None:
            warn.append(f"zone « {zid} » : dimensions incomplètes, elle ne compte "
                        "dans aucun total de volume")

    connus = vocabulaire(rayons)
    # Qui est rangé où : c'est cette table qui permet de voir deux denrées
    # incompatibles partager une étagère, ce qu'aucune ligne ne dit toute seule.
    par_zone = {}

    for i, d in enumerate(gm["denrees"]):
        cid = d.get("ingredient", "?")
        ou = f"denrée « {cid} »"
        zid = d.get("zone")
        if zid not in zones:
            err.append(f"{ou} : zone « {zid} » inconnue")
        else:
            par_zone.setdefault(zid, []).append(d)
        if cid not in connus:
            err.append(f"{ou} : ingrédient sans rayon (à ajouter dans rayons.yaml)")
        if not d.get("unites"):
            err.append(f"{ou} : `unites` doit valoir au moins 1 — une denrée qu'on "
                       "n'a pas ne se range pas, elle s'efface")
        if d.get("etat") not in ETATS:
            err.append(f"{ou} : état « {d.get('etat')} » inconnu — attendu parmi {list(ETATS)}")
        if d.get("nature", "autre") not in NATURES:
            err.append(f"{ou} : nature « {d.get('nature')} » inconnue — attendu parmi {list(NATURES)}")
        if d.get("acidite", ACIDITE_DEFAUT) not in ("basse", "haute"):
            err.append(f"{ou} : acidité « {d.get('acidite')} » inconnue — attendu `basse` ou `haute`")
        for s in d.get("sensible") or []:
            if s not in AGRESSIONS:
                err.append(f"{ou} : sensibilité « {s} » inconnue "
                           f"— attendu parmi {sorted(AGRESSIONS)}")

    # Puis les erreurs de rangement, calculées au même endroit que celles que
    # l'export publie : deux implémentations de « craint la lumière » finiraient
    # par ne plus dire la même chose.
    warn += alertes_rangement(gm)
    return err, warn

File: src/model/gardeManger.ts
// Ce que le garde-manger apporte au CHOIX D'UN PLAT.
//
// `stock.vue.ts` montre le garde-manger ; ici on s'en sert. La question n'est
// plus « qu'est-ce qu'il y a dans ce placard » mais « qu'est-ce qui va se
// perdre, et quel plat le mangerait ».
//
// DEUX RÉCOMPENSES DISTINCTES, ET LES CONFONDRE CASSERAIT LES DEUX.
//
// *Utiliser* le placard est déjà payé, et ailleurs : un ingrédient du
// garde-manger ne crée pas de ligne de courses, donc `article_marginal` ne monte
// pas. Un plat bâti sur ce qu'on a gagne déjà, sans que personne ait eu à
// l'écrire.
//
// CETTE PHRASE A ÉTÉ FAUSSE ENTRE T22 ET T24, ce qui vaut d'être noté parce
// qu'elle servait à justifier tout le reste. Elle n'était vraie que de
// `rayons.placard` — le sel, l'huile. Les 45 ingrédients du garde-manger, eux,
// étaient rangés dans `rayons.épicerie` : `provenance()` les disait `courses`,
// ils produisaient une ligne d'achat, et un plat qui puisait dans le stock était
// donc PÉNALISÉ de 0,4 au lieu d'être neutre. La conclusion tenait, la raison
// était fausse. T24 a branché `provenance()` sur le relevé, et la phrase est
// redevenue vraie.
//
// *Sauver* le placard est ce qui manquait. Une boîte de maïs tient trois ans ;
// un sachet d'épeautre ouvert rancit en six mois ; quatre kilos de pommes de
// terre au-dessus d'un siphon germent cette semaine. Seules les deux dernières
// méritent qu'on déplace un dîner, et c'est pour elles seules que le score paie.
//
// La conséquence tient en une ligne : **on ne paie que ce qui est à risque.**
// Payer aussi les conserves ferait gagner les plats à longue liste d'épicerie,
// ce qui est exactement l'inverse du service rendu.
//
// ────────────────────────────────────────────────────────────────────────────
// CE QUE CE MODÈLE NE SAIT PAS FAIRE, ET IL FAUT LE SAVOIR EN LE LISANT.
//
// Un aromate n'est pas sauvé parce qu'un plat le cite. Mesuré sur le corpus :
// l'oignon paraît dans 42 % des 86 plats, l'ail dans 19 %. Ces deux-là seront
// mangés de toute façon — ils n'ont besoin d'aucun coup de pouce, et le terme
// qui les signale est presque une constante, donc il ne départage rien. Ce qui
// se perd vraiment chez ce foyer, ce sont les six kilos de pommes de terre, les
// deux farines d'épeautre ouvertes et les quatre fonds de paquets de pâtes.
//
// Distinguer les deux demanderait de comparer CE QU'ON A à CE QU'UN PLAT PREND
// — un oignon sur un filet, contre 800 g de pommes de terre sur deux kilos. Le
// relevé ne porte pas ces quantités pour le frais (`par_unite: null`), et les
// inventer donnerait un classement qui a l'air fin et ne l'est pas.
//
// D'où le partage assumé : le score NUDGE, et `aSauver()` DÉSIGNE. La liste,
// elle, nomme les pommes de terre et l'épeautre sans se laisser noyer par les
// aromates. C'est elle qu'il faut lire pour savoir quoi cuisiner ce soir.
// ────────────────────────────────────────────────────────────────────────────

import { fractionDe, marque } from "./axe";
import type { Catalogue, ConservationDenree, Denree, Etat, Plat, Urgence } from "./types";

/** Du plus pressé au moins. Ne sert plus qu'à ORDONNER la liste « à sauver » :
 *  le score, lui, est passé sur l'axe 0–1 de `axe.ts` — deux échelles pour deux
 *  usages seraient deux occasions de diverger, mais celle-ci ne classe que des
 *  lignes d'écran et n'entre dans aucun calcul. */
const RANG: Record<Urgence, number> = { haute: 3, moyenne: 2, basse: 1 };

/** L'id d'achat, alias résolus. Le garde-manger et les recettes puisent dans le
 *  même vocabulaire, mais rien ne garantit qu'ils aient choisi la même
 *  orthographe — `oignons` d'un côté, `oignon` de l'autre. Sans ce passage, le
 *  rapprochement échoue en silence et le bonus ne tombe jamais. */
const alias = (catalogue: Catalogue, id: string): string => catalogue.rayons.aliases[id] ?? id;

/**
 * CE QUE LE SCORE PAIE, PAR ÉTAT DE LA DENRÉE — T60.
 *
 * Une seule ligne, et c'est la règle entière : **le score ne paie que la
 * barrière rompue.** Un paquet ouvert a une horloge que cuisiner arrête, et
 * c'est la seule chose vraie que le relevé sache dire — `urgence()` le dit déjà
 * mot pour mot côté Python : « le paquet entamé : la barrière est rompue,
 * l'horloge tourne ».
 *
 * CE QUE ÇA ÉCARTE, ET POURQUOI CE N'EST PAS LE CHEMIN QUE LE TICKET AVAIT PRÉVU.
 * T60 demandait d'écarter l'artefact du sous-évier en retirant du score les
 * denrées que leur ZONE abîme. Mesuré, ce geste-là ne fait rien : les quatre
 * légumes sous l'évier sont `etat: frais`, et `urgence()` rend `haute` sur le
 * frais AVANT même de regarder la zone. Ils seraient tout aussi `haute` dans une
 * cave sèche. Le seul `haute` purement dû à sa zone est le sachet de pignons —
 * qui n'est dans aucun plat, donc n'a jamais rien payé. Retirer « la cause
 * zone » aurait laissé le +5 sur l'oignon intact et le ticket clos à tort.
 *
 * CE QUI EST VRAI, ET QUE CE FICHIER ÉCRIVAIT DÉJÀ EN TÊTE : l'oignon est dans
 * 42 % des 86 plats, l'ail dans 19 %. Ces quatre-là seront mangés de toute
 * façon ; le terme qui les signale est presque une constante, donc il ne
 * départage rien — il déplace juste tout le classement vers les plats qui
 * contiennent un aromate. Le frais du primeur n'est d'ailleurs pas une denrée de
 * placard qu'on sauve : c'est la classe que T46 avait déjà dû clôturer pour les
 * planchers, sur exactement les mêmes quatre ids (ail, échalote, oignon, pomme
 * de terre).
 *
 * L'APP N'ARRÊTE PAS DE LE DIRE POUR AUTANT, ELLE ARRÊTE DE LE PAYER. Le frais
 * reste dans « À manger en premier », qui est faite pour être lue ; le mauvais
 * rangement reste un geste (« pomme de terre, oignon… — dans un endroit humide
 * (sous-évier) »), qui existe depuis T31. C'est le partage que ce fichier
 * annonce en tête : le score NUDGE, `aSauver()` DÉSIGNE.
 *
 * ⚠ CE QUE ÇA NE FERME PAS. Le vrai axe n'est pas l'état, c'est l'ubiquité — un
 * terme qui tire sur 42 % du corpus ne départage rien, quel que soit son motif.
 * Le mesurer demanderait de comparer CE QU'ON A à CE QU'UN PLAT PREND, et le
 * relevé ne porte pas les quantités du frais (`par_unite: null`). L'état est le
 * meilleur proxy disponible ; il coûte une courgette fraîche qui, elle, mériterait
 * peut-être d'être payée. On la perdra sans le voir jusqu'à ce que les quantités
 * existent.
 */
const ECOULEMENT: Partial<Record<Etat, Urgence>> = { entame: "moyenne" };

/**
 * Ce que le placard offre à écouler, par ingrédient, sur l'axe 0–1 de T57.
 *
 * UNE ENTRÉE PAR INGRÉDIENT, PAS PAR LOT. Le score se pose sur une ligne de
 * recette, qui nomme un ingrédient ; compter quatre fois `pates` parce que
 * quatre fonds de paquet traînent ferait d'un plat de pâtes le meilleur dîner de
 * la semaine, quatre fois de suite.
 *
 * ON PROJETTE L'ÉTAT, PAS L'URGENCE ENREGISTRÉE. Un paquet entamé rangé dans une
 * mauvaise zone sort `haute` de l'export ; il vaut quand même 0,4 ici, parce que
 * ce que cuisiner rattrape est l'ouverture du paquet, pas l'humidité de
 * l'étagère. La différence ne se voit pas sur le relevé du 26/08 — aucune denrée
 * entamée n'est mal rangée — et c'est justement pourquoi elle est écrite.
 *
 * UNE DENRÉE QUE NUL PLAT NE CONSOMME N'Y ARRIVE JAMAIS, sans qu'on ait à la
 * filtrer : `bonusPlacard` part des lignes du PLAT et cherche dedans, jamais
 * l'inverse. Les quatre `moyenne` hors recette du relevé — cracotte, krisprolls,
 * blé-lentilles, farine d'épeautre — ne peuvent donc rien bruiter, et le volet
 * (a) de T60 était déjà tenu par la forme de la boucle. Un test le tient
 * maintenant explicitement, parce qu'une propriété vraie par accident se perd au
 * premier refactor.
 */
export function aEcouler(catalogue: Catalogue): Map<string, number> {
  const par = new Map<string, number>();
  for (const d of catalogue.gardeManger.denrees) {
    const u = ECOULEMENT[d.etat];
    if (!u) continue;
    const f = fractionDe(u);
    if (f == null) continue;
    const id = alias(catalogue, d.ingredient);
    par.set(id, Math.max(par.get(id) ?? 0, f));
  }
  return par;
}

/** Au-delà, le plat ne gagne plus rien — T59. Trois articles sauvés valent trois
 *  fois un article ; le quatrième ne dit plus rien de neuf sur ce plat-là, il
 *  dit qu'il a une longue liste d'ingrédients, ce que `article_marginal` fait
 *  déjà payer. Le plafond n'est PAS une dégressivité : les trois premiers
 *  comptent plein, ce qui était la demande explicite du ticket. */
export const PLAFOND_ARTICLES = 3;

export interface BonusPlacard {
  /** Ce que ça vaut au score. Zéro quand le plat ne sauve rien. */
  score: number;
  /** Les ingrédients sauvés, tels qu'on les dira à l'écran. */
  noms: string[];
  /** Y en a-t-il un qui est vraiment pressé ? Décide de la phrase. */
  urgent: boolean;
}

/**
 * Ce qu'un plat rattrape du placard.
 *
 * ON COMPTE DES INGRÉDIENTS DISTINCTS, JAMAIS DES LIGNES. Une recette peut citer
 * l'oignon deux fois — dans la garniture et dans la sauce — et le payer deux
 * fois donnerait un avantage à la façon dont la recette est écrite plutôt qu'à
 * ce qu'elle mange.
 *
 * Les lignes `from_accepts` sont ignorées : elles réclament une base cuisinée
 * (« 250 g de lentilles cuites »), pas une matière première, et le chaînage a
 * déjà ses propres poids pour ça.
 */
export function bonusPlacard(
  catalogue: Catalogue,
  plat: Plat,
  pressees: ReadonlyMap<string, number>,
  poids: Record<string, number>,
): BonusPlacard {
  const vus = new Map<string, number>();
  for (const ing of plat.ingredients) {
    if (ing.base) continue;
    const id = alias(catalogue, ing.id);
    const f = pressees.get(id);
    // Absent de l'axe = rien à écouler ici : une conserve scellée attendra, un
    // légume frais se règle par le rangement et par la liste. Voir `ECOULEMENT`.
    if (f == null) continue;
    vus.set(id, f);
  }

  // LE SCORE CUMULE ENFIN, ET C'EST L'ABANDON D'UNE RÈGLE ÉCRITE ICI — T59.
  //
  // La version d'avant payait un forfait par plat, jamais par ingrédient, parce
  // que cumuler donnait +10 à presque tout ce qui contient un oignon et de l'ail
  // et faisait gagner les plats à LONGUE LISTE D'INGRÉDIENTS. L'argument était
  // juste, le remède visait à côté : ce qui bruitait n'était pas le cumul, c'était
  // que le frais ubiquitaire soit payé du tout. T60 le retire à la source, et le
  // cumul redevient ce qu'il aurait toujours dû être — un plat qui sauve trois
  // choses vaut mieux qu'un plat qui en sauve une.
  //
  // L'OBJECTION D'ÉQUILIBRAGE A ÉTÉ SOULEVÉE ET ÉCARTÉE, et elle vaut d'être
  // relue avant d'être ressortie : un plafond à +15 dépasse `proteine_manquante:
  // 6`. Trois raisons. Le levier de correction est le nombre `ecoule` lui-même et
  // non un mécanisme de plus (`equilibre.yaml` : « valeur posée à vue, à régler à
  // l'usage »). Le mode d'échec redouté — la répétition — est déjà défendu trois
  // fois : `repetition_profil: -4`, `repetition_feculent: -3`, `cooldown_jours:
  // 10` depuis T52. Et surtout LA DOMINATION EST LE CAHIER DES CHARGES :
  // Workspace#41 demande d'encourager « au maximum » l'utilisation des stocks,
  // donc l'écoulement qui passe devant la protéine manquante n'est pas un
  // déséquilibre, c'est l'app qui fait son travail.
  //
  // LES PLUS PRESSÉS D'ABORD, quand il y en a plus de trois : le plafond doit
  // couper la queue de la liste, pas un article au hasard de l'ordre des lignes
  // de la recette. À égalité on tranche par nom, pour que deux rendus de la même
  // semaine donnent le même score.
  const classes = [...vus].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0], "fr"));
  const payes = classes.slice(0, PLAFOND_ARTICLES);
  const score = (poids["ecoule"] ?? 0) * payes.reduce((s, [, f]) => s + f, 0);

  return {
    // Arrondi au dixième comme le score entier : 5 × 0,4 doit donner 2 et non
    // 2,0000000000000004, sans quoi un test de non-régression du placard
    // deviendrait un test de virgule flottante.
    score: Math.round(score * 10) / 10,
    // TOUS LES SAUVÉS SONT NOMMÉS, MÊME AU-DELÀ DU PLAFOND. La phrase dit ce que
    // le plat sauve, le score dit ce que ça vaut ; n'en nommer que trois ferait
    // mentir la première pour justifier le second.
    noms: classes.map(([id]) => id.replace(/-/g, " ")),
    // La même échelle décide du chiffre ET du mot : `urgent` est vrai quand le
    // plus pressé des articles a franchi le seuil haut de T57. Sur le relevé du
    // 26/08 il ne l'est jamais — le placard n'offre plus que des paquets entamés,
    // à 0,4 — et la phrase urgente attend le dépôt, que T47 branchera ici.
    urgent: marque(classes[0]?.[1] ?? null) === "urgent",
  };
}

export interface ASauver {
  ingredient: string;
  nom: string;
  urgence: Urgence;
  /** Où c'est rangé, en toutes lettres — on ne sauve pas ce qu'on ne trouve pas. */
  zone: string;
  /** Pourquoi ça presse, dit en français. */
  raison: string;
  /**
   * L'AUTRE ISSUE : transformer au lieu de cuisiner.
   *
   * Une denrée qui court a deux sorties, et le modèle les connaît toutes les
   * deux depuis le prototype. La première est de la manger ce soir — c'est ce
   * que le score pousse. La seconde est d'arrêter son horloge, et c'est souvent
   * la bonne : on ne mange pas six kilos de pommes de terre parce qu'ils
   * germent.
   */
  conserver: ConservationDenree[];
  /** Ce qu'il faudrait acquérir pour avoir une issue de plus. Le nœud de
   *  compétence, pas une suggestion d'achat — la distinction vient de #29. */
  verrouille: ConservationDenree[];
}

/**
 * Ce qui se perd, du plus pressé au moins.
 *
 * L'ÉCRAN NE PEUT PAS SE CONTENTER DE L'URGENCE. « pommes de terre : haute » ne
 * dit pas quoi faire. On rend la RAISON — frais, entamé, ou abîmé par sa zone —
 * parce que c'est elle qui distingue « mange-les » de « range-les ailleurs ».
 */
export function aSauver(catalogue: Catalogue): ASauver[] {
  const zones = new Map(catalogue.gardeManger.zones.map((z) => [z.id, z]));
  const raison = (d: Denree): string => {
    const z = zones.get(d.zone);
    if (z) {
      if (d.sensible.includes("lumiere") && z.exposition === "jour") return "s’abîme à la lumière";
      if (d.sensible.includes("humidite") && z.hygrometrie === "humide") return "s’abîme à l’humidité";
      if (d.sensible.includes("chaleur") && z.chaleur) return "s’abîme à la chaleur";
    }
    if (d.etat === "frais") return "frais, ne se garde pas";
    return "paquet entamé";
  };
  return catalogue.gardeManger.denrees
    .filter((d) => d.urgence !== "basse")
    .map((d) => ({
      ingredient: d.ingredient,
      nom: d.ingredient.replace(/-/g, " "),
      urgence: d.urgence,
      zone: zones.get(d.zone)?.label ?? d.zone,
      raison: raison(d),
      conserver: d.conservations.filter((c) => c.acquis),
      verrouille: d.conservations.filter((c) => !c.acquis),
    }))
    .sort((a, b) => RANG[b.urgence] - RANG[a.urgence] || a.nom.localeCompare(b.nom, "fr"));
}
